### DISCUSSION TITLE
Stand-alone example of proxy-wasm-cpp-host

### DISCUSSION THREAD

**Discussion by bhakta0007** [proxy-wasm/proxy-wasm-cpp-host#346] [Q&A]
Looking for a stand-alone where this module is compiled to a working program demonstrating a simple working use-case. Is there one such example available?

I have seen the usage in projects like ATS - looking for a simple starter to understand the interactions.

**Comment by bhakta0007** [comment]
searching on Chat GPT , I got this reference

https://github.com/proxy-wasm/proxy-wasm-cpp-host/tree/main/examples/http-filter

and this usage example

```
$ mkdir build
$ cd build
$ cmake ..
$ make
$ proxy-wasm-cli run --root . --config http-filter.json http-filter.wasm
```

I cannot find this in the tree. Is the "proxy-wasm-cli" available someplace else?

**Reply by PiotrSikora** [reply]
> searching on Chat GPT , I got this reference
> 
> https://github.com/proxy-wasm/proxy-wasm-cpp-host/tree/main/examples/http-filter
> 
> and this usage example
> 
> ```
> $ mkdir build
> $ cd build
> $ cmake ..
> $ make
> $ proxy-wasm-cli run --root . --config http-filter.json http-filter.wasm
> ```
> 
> I cannot find this in the tree. Is the "proxy-wasm-cli" available someplace else?

Haha, that response is completely fabricated:

1) `proxy-wasm-cpp-host` uses the old-style `master` branch and not `main` branch.
2) There was never an `examples` directory.
3) `proxy-wasm-cpp-host` uses Bazel and not CMake, so those build instructions are wrong.
4) There was never a `proxy-wasm-cli` program.

Basically, every single line was a lie.

**Reply by bhakta0007** [reply]
I really fell for the answer and spent time searching around. It appears that AI can also get cross-linked memories, just like human mind.  ;-)

However, the proxy-wasm-cli is definitely a good inspiration!

**Comment by martijneken** [comment]
Hey there, I think one problem you'll run into is that the host is designed to be coupled to a network proxy. In essence ProxyWasm runs Wasm modules in response to events on a network proxy (e.g. on request, on response), making it hard to use "stand-alone". That said, there are a number of examples in the unit tests (test/ folder), and more examples for Wasm implementers are being developed.

**Comment by bhakta0007** [comment]
Thanks for the response @martijneken . I understand that this is designed to be coupled with a network proxy. I was trying to see if theres is an example usage where I don't have to dig into a specific proxy code-base. One could still write a simple wrapper with a main (more like a cli) which can make some calls into the VM and dumps the results - that would be a good way to demonstrate the capabilities of this project.

**Reply by PiotrSikora** [reply]
> Thanks for the response @martijneken . I understand that this is designed to be coupled with a network proxy. I was trying to see if theres is an example usage where I don't have to dig into a specific proxy code-base. One could still write a simple wrapper with a main (more like a cli) which can make some calls into the VM and dumps the results - that would be a good way to demonstrate the capabilities of this project.

We've discussed something like that in the past, mostly as a way to exercise various code paths in tests in this repo, instead of relying on PRs to proxies integrating with it.

I think a CLI tool that takes request/response from the stdin, file or a command line would be nice to have, but someone needs to write it.

**Reply by bhakta0007** [reply]
@PiotrSikora Other than looking into some proxy code, what is the best place to understand the APIs and the usage. While I ask this, I am going through ATS code and I see I need to sift through a lot of proxy specific logic to understand the typical sequence of events in order to:

1. Load a VM
2. Call a function in the vm (e.g. onRequestHeaders)
3. Print back the results

**Reply by shukitchan** [reply]
> @PiotrSikora Other than looking into some proxy code, what is the best place to understand the APIs and the usage. While I ask this, I am going through ATS code and I see I need to sift through a lot of proxy specific logic to understand the typical sequence of events in order to:
> 
> 1. Load a VM
> 2. Call a function in the vm (e.g. onRequestHeaders)
> 3. Print back the results

Also please feel free to ask questions by creating issues in ATS github project.

**Comment by martijneken** [comment]
I just sent this unit test PR to hopefully give you a starting point. It was useful for me as well. https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/348

Note that this doesn't have any HTTP header logic. For that you'll need to implement a Stream Context that implements HeaderInterface: https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/491915a9a999a18e394af22f9103b4d40ed0d24c/include/proxy-wasm/context_interface.h#L341

**Comment by bhakta0007** [comment]
thanks everyone. This is my first time doing C++ (coming from the C world)  - which makes it a bit harder to look at these projects. With @martijneken 's test-case pointer (thanks!), I was able to get a call going into the WASM, so I got past the first step.

I plan to continue working on this further. I will post further questions/updates as I make progress.

**Comment by bhakta0007** [comment]
Update: I have a fork https://github.com/bhakta0007/proxy-wasm-cpp-host

I am working on a stand-alone utility on top of this project. My goal of building this was to understand how it works. I took most of the code from Apache Traffic Server project.

If you guys find this interesting and useful, I can send a merge request to this project. [Of course the code needs a lot of cleanup!]
